src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1020 @@
+agg_data_is_newer(struct ec_agg_entry *entry)
+{
+	struct ec_agg_extent	*agg_extent;
+
+	d_list_for_each_entry(agg_extent, &entry->ae_cur_stripe.as_dextents,
+			      ae_link) {
+		if (agg_extent->ae_epoch < entry->ae_par_extent.ape_epoch)
+			return false;
+	}
+	return true;
+}
+
+/* Determines if the replicas present for the current stripe of object entry
+ * constitute a full stripe. If parity exists for the the stripe, the replicas
+ * making up the full stripe must be a later epoch that the parity.
+ */
+static bool
+agg_stripe_is_filled(struct ec_agg_entry *entry, bool has_parity)
+{
+	bool	is_filled, rc = false;
+
+	D_ASSERT(entry->ae_cur_stripe.as_stripe_fill <=
+		 obj_ec_stripe_rec_nr(entry->ae_oca));
+
+	is_filled = entry->ae_cur_stripe.as_stripe_fill ==
+		obj_ec_stripe_rec_nr(entry->ae_oca);
+
+	if (is_filled)
+		if (!has_parity || agg_data_is_newer(entry))
+			rc = true;
+	return rc;
+}
+
+/* Range to not delete, if the extent carries over into the next stripe.
+ */
+static unsigned int
+agg_get_carry_under(struct ec_agg_entry *entry)
+{
+	struct ec_agg_extent *agg_extent, *ext_tmp;
+	unsigned int	      tail = 0U;
+
+	d_list_for_each_entry_safe(agg_extent, ext_tmp,
+				   &entry->ae_cur_stripe.as_dextents,
+				   ae_link) {
+		/* Check for -over extent. */
+		tail = agg_carry_over(entry, agg_extent);
+		/* At most one extent should carry over. */
+		if (tail)
+			return agg_extent->ae_recx.rx_nr - tail;
+		/* At most one extent should carry over. */
+	}
+	return 0;
+}
+
+/* Writes updated parity to VOS, and removes replicas fully contained
+ * in the processed stripe.
+ */
+static int
+agg_update_vos(struct ec_agg_entry *entry, bool write_parity)
+{
+	daos_epoch_range_t	 epoch_range = { 0 };
+	daos_recx_t		 recx = { 0 };
+	struct ec_agg_param	*agg_param;
+	unsigned int		 len = entry->ae_oca->u.ec.e_len;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	int			 rc = 0;
+
+	agg_param = container_of(entry, struct ec_agg_param, ap_agg_entry);
+
+	if (write_parity) {
+		d_sg_list_t	 sgl = { 0 };
+		daos_iod_t	 iod = { 0 };
+		d_iov_t		*iov;
+
+		iov = &entry->ae_sgl.sg_iovs[AGG_IOV_PARITY];
+		sgl.sg_iovs = iov;
+		sgl.sg_nr = 1;
+		iod.iod_nr = 1;
+		iod.iod_size = entry->ae_rsize;
+		iod.iod_name = entry->ae_akey;
+		iod.iod_type = DAOS_IOD_ARRAY;
+		iod.iod_recxs = &recx;
+		recx.rx_idx = (entry->ae_cur_stripe.as_stripenum * len) |
+							PARITY_INDICATOR;
+		recx.rx_nr = len;
+		rc = vos_obj_update(agg_param->ap_cont_handle, entry->ae_oid,
+				    entry->ae_cur_stripe.as_hi_epoch, 0, 0,
+				    &entry->ae_dkey, 1, &iod, NULL, &sgl);
+		if (rc) {
+			D_ERROR("vos_obj_update failed: "DF_RC"\n", DP_RC(rc));
+			goto out;
+		}
+	}
+	recx.rx_idx = entry->ae_cur_stripe.as_stripenum * k * len -
+		entry->ae_cur_stripe.as_prefix_ext;
+	recx.rx_nr = k * len + entry->ae_cur_stripe.as_prefix_ext -
+		entry->ae_cur_stripe.as_suffix_ext;
+	epoch_range.epr_lo = agg_param->ap_epr.epr_lo;
+	epoch_range.epr_hi = entry->ae_cur_stripe.as_hi_epoch;
```

Comment:
The highest extent epoch in the stripe is correct for the remove.  In all cases except one, the parity has been written (code above), at the highest extent epoch in the stripe.  In the case where parity isn't updated, it's because all of the extents in the stripe are older than the existing parity, so this upper bound is still correct.

Reply:
Any response to this? Am I understanding correct?